File: netlify/functions/recipes.py
"""
Netlify Function: Recipes API
Handles GET (list all recipes) and POST (create new recipe)
"""

import json
from utils.db import execute_insert, execute_query, format_response, handle_error


def handler(event, context):
    """Main handler for recipes endpoint"""

    # Handle CORS preflight
    if event["httpMethod"] == "OPTIONS":
        return format_response(200, {})

    try:
        # GET: List all recipes
        if event["httpMethod"] == "GET":
            return get_recipes(event)

        # POST: Create new recipe
        elif event["httpMethod"] == "POST":
            return create_recipe(event)

        else:
            return format_response(
                405, {"error": "Method not allowed", "success": False}
            )

    except Exception as e:
        return handle_error(e)


def get_recipes(event):
    """Get all recipes with optional filtering"""
    try:
        # Parse query parameters
        params = event.get("queryStringParameters") or {}
        search = params.get("search", "").strip()
        category = params.get("category", "").strip()

        # Validate and sanitize limit/offset
        try:
            limit = min(int(params.get("limit", 100)), 1000)  # Cap at 1000
            offset = max(int(params.get("offset", 0)), 0)  # No negative offsets
        except ValueError:
            return format_response(
                400, {"error": "Invalid pagination parameters", "success": False}
            )

        # Build query with proper parameterization
        query = "SELECT * FROM recipes WHERE 1=1"
        query_params = []

        # Apply search filter (parameterized to prevent SQL injection)
        if search:
            query += " AND (titre ILIKE %s OR instructions::text ILIKE %s)"
            search_pattern = f"%{search}%"
            query_params.extend([search_pattern, search_pattern])

        # Apply category filter
        if category:
            query += " AND category = %s"
            query_params.append(category)

        # Order by creation date (newest first)
        query += " ORDER BY created_at DESC"

        # Apply pagination
        query += " LIMIT %s OFFSET %s"
        query_params.extend([limit, offset])

        # Execute query
        recipes = execute_query(query, tuple(query_params), fetch="all")

        return format_response(
            200, {"success": True, "recipes": recipes, "count": len(recipes)}
        )

    except Exception as e:
        return handle_error(e)


def create_recipe(event):
    """Create a new recipe"""
    try:
        # Parse request body
        body = json.loads(event["body"])

        # Validate required fields
        required_fields = ["titre", "ingredients", "instructions"]
        for field in required_fields:
            if field not in body:
                return format_response(
                    400, {"error": f"Missing required field: {field}", "success": False}
                )

        # Validate data types
        if not isinstance(body["ingredients"], list):
            return format_response(
                400, {"error": "ingredients must be an array", "success": False}
            )

        if not isinstance(body["instructions"], (list, str)):
            return format_response(
                400,
                {"error": "instructions must be an array or string", "success": False},
            )

        # Check for duplicate recipe title
        check_query = "SELECT id FROM recipes WHERE titre = %s"
        existing = execute_query(check_query, (body["titre"],), fetch="one")
        if existing:
            return format_response(
                409,
                {"error": "Une recette avec ce titre existe déjà", "success": False},
            )

        # Prepare recipe data
        insert_query = """
            INSERT INTO recipes (
                titre, temps_preparation, temps_cuisson, rendement,
                ingredients, instructions, image_url, category, tags
            )
            VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s)
            RETURNING *
        """

        recipe_params = (
            body["titre"],
            body.get("temps_preparation", ""),
            body.get("temps_cuisson", ""),
            body.get("rendement", ""),
            json.dumps(body["ingredients"]),
            json.dumps(body["instructions"]),
            body.get("image_url", ""),
            body.get("category", ""),
            body.get("tags", []),
        )

        # Insert recipe
        recipe = execute_insert(insert_query, recipe_params)

        return format_response(
            201,
            {"success": True, "recipe": recipe, "message": "Recette créée avec succès"},
        )

    except json.JSONDecodeError:
        return format_response(
            400, {"error": "Invalid JSON in request body", "success": False}
        )
    except Exception as e:
        return handle_error(e)

File: netlify/functions/recipes.mjs
import { json, err } from "./shared/http.mjs";
import { queryAll, queryOne, execReturning } from "./shared/db.mjs";

function parseParams(qs) {
  if (!qs) return {};
  const out = {};
  for (const [k, v] of Object.entries(qs)) {
    if (v != null) out[k] = v;
  }
  return out;
}

export const handler = async (event) => {
  if (event.httpMethod === "OPTIONS") return json(200, {});

  try {
    if (event.httpMethod === "GET") {
      const params = parseParams(event.queryStringParameters || {});
      const search = (params.search || "").trim();
      const category = (params.category || "").trim();
      let limit = 100;
      let offset = 0;
      try {
        limit = Math.min(parseInt(params.limit ?? "100", 10) || 100, 1000);
        offset = Math.max(parseInt(params.offset ?? "0", 10) || 0, 0);
      } catch {
        return err(400, "Invalid pagination parameters");
      }

      const parts = ["SELECT * FROM recipes WHERE 1=1"];
      const args = [];
      if (search) {
        const p = `%${search}%`;
        args.push(p);
        parts.push(`AND (titre ILIKE $${args.length} OR instructions::text ILIKE $${args.length})`);
      }
      if (category) {
        args.push(category);
        parts.push(`AND category = $${args.length}`);
      }
      parts.push("ORDER BY created_at DESC");
      args.push(limit);
      const li = args.length;
      args.push(offset);
      const oi = args.length;
      parts.push(`LIMIT $${li} OFFSET $${oi}`);

      const sql = parts.join(" ");
      const recipes = await queryAll(sql, args);
      return json(200, { success: true, recipes, count: recipes.length });
    }

    if (event.httpMethod === "POST") {
      let body;
      try {
        body = JSON.parse(event.body || "{}");
      } catch {
        return err(400, "Invalid JSON in request body");
      }

      const required = ["titre", "ingredients", "instructions"];
      for (const f of required) {
        if (!(f in body)) return err(400, `Missing required field: ${f}`);
      }
      if (!Array.isArray(body.ingredients)) {
        return err(400, "ingredients must be an array");
      }
      if (typeof body.instructions !== "string" && !Array.isArray(body.instructions)) {
        return err(400, "instructions must be an array or string");
      }

      const dup = await queryOne("SELECT id FROM recipes WHERE titre = $1", [body.titre]);
      if (dup) {
        return err(409, "Une recette avec ce titre existe déjà");
      }

      const row = await execReturning(
        `INSERT INTO recipes (
          titre, temps_preparation, temps_cuisson, rendement,
          ingredients, instructions, image_url, category, tags, source_url
        )
        VALUES ($1, $2, $3, $4, $5::jsonb, $6::jsonb, $7, $8, $9, $10)
        RETURNING *`,
        [
          body.titre,
          body.temps_preparation ?? "",
          body.temps_cuisson ?? "",
          body.rendement ?? "",
          JSON.stringify(body.ingredients),
          JSON.stringify(body.instructions),
          body.image_url ?? "",
          body.category ?? "",
          body.tags ?? [],
          body.source_url ?? "",
        ]
      );

      return json(201, {
        success: true,
        recipe: row,
        message: "Recette créée avec succès",
      });
    }

    return err(405, "Method not allowed");
  } catch (e) {
    console.error(e);
    return json(500, { success: false, error: "An internal error occurred" });
  }
};
